src/prefect/tasks/kubernetes/job.py

```diff
@@ -695,34 +698,68 @@ def run(
                 "The job name must be defined in the body under the metadata key."
             )
 
-        api_client = cast(
+        api_client_job = cast(
             client.BatchV1Api, get_kubernetes_client("job", kubernetes_api_key_secret)
         )
 
-        api_client.create_namespaced_job(namespace=namespace, body=body, **kube_kwargs)
+        api_client_pod = cast(
+            client.CoreV1Api, get_kubernetes_client("pod", kubernetes_api_key_secret)
+        )
+
+        api_client_job.create_namespaced_job(
+            namespace=namespace, body=body, **kube_kwargs
+        )
         self.logger.info(f"Job {job_name} has been created.")
 
-        completed = False
-        while not completed:
-            res = api_client.read_namespaced_job_status(
-                name=job_name, namespace=namespace
+        job = api_client_job.read_namespaced_job_status(
+            name=job_name, namespace=namespace
+        )
+
+        pod_selector = f"controller-uid={job.metadata.labels['controller-uid']}"
+        pods_list = api_client_pod.list_namespaced_pod(
+            namespace=namespace, label_selector=pod_selector
+        )
+
+        if len(pods_list.items) != 1:
+            raise signals.FAIL(
+                f"Job {job_name} failed to match pod with controller-uid {pod_selector}. "
+                + f"Expected one, got {len(pods_list.items)}."
             )
 
-            if res.status.active:
-                time.sleep(job_status_pool_interval)
-            else:
-                if res.status.failed:
-                    raise signals.FAIL(
-                        f"Job {job_name} failed, check Kubernetes pod logs for more information."
-                    )
-                elif res.status.succeeded:
-                    self.logger.info(f"Job {job_name} has been completed.")
-                    completed = True
-
-        if delete_job_after_completion:
-            api_client.delete_namespaced_job(
-                name=job_name,
-                namespace=namespace,
-                body=client.V1DeleteOptions(propagation_policy="Foreground"),
+        pod = pods_list.items[0]
+        pod_name = pod.metadata.name
+        self.logger.info(pod)
+        while pod.status.phase != "Running":
+            time.sleep(1)
+            pod = api_client_pod.read_namespaced_pod(name=pod_name, namespace=namespace)
+
+        self.logger.info(f"Pod {pod_name} is running.")
+
+        # The watch stream is automatically stopped when the job is completed or failed. So we don't
+        # have to poll the job its status.
```

Comment:
Watch streams can complete for other reasons (stale event requiring a restart, api error, etc...). Most code I've used them required an extra loop around `stream` calls to handle restarts/spurious errors.